src/sentry/hybridcloud/models/outbox.py

```diff
@@ -224,60 +227,206 @@ def process_coalesced(
+                if coalesced is None:
+                    yield None
+                    return
+
+                # For timing and metrics, we also need to determine the first (oldest) record.
+                # This is done with a second query ordering by ascending id. If no record is found,
+                # fall back to the representative record.
+                first_coalesced: OutboxBase | None = (
+                    self.select_coalesced_messages()
+                    .select_for_update(nowait=True)
+                    .order_by("id")
+                    .first()
+                ) or coalesced
+
+                # Build a tags dictionary for metrics purposes. The outbox category is included
+                # to help report timing metrics by type and to show if the flush was synchronous.
+                tags: dict[str, int | str] = {
+                    "category": OutboxCategory(self.category).name,
+                    "synchronous": int(is_synchronous_flush),
+                }
+
+                # Assert that we have a valid first record (which should always be true for a non-empty group).
+                assert (
+                    first_coalesced
+                ), "first_coalesced incorrectly set for non-empty coalesce group"
+
+                # Record the net queue time metric using the date the first message was added.
+                metrics.timing(
+                    "outbox.coalesced_net_queue_time",
+                    datetime.datetime.now(tz=datetime.UTC).timestamp()
+                    - first_coalesced.date_added.timestamp(),
+                    tags=tags,
+                )
+
+                # Store the representative (highest id) and timing values for later use.
+                coalesced_id = coalesced.id
+                # 'scheduled_from' is used for metrics and can serve as the original scheduled timestamp.
+                scheduled_from = first_coalesced.scheduled_from
+                date_added = first_coalesced.date_added
+
+                # Get a list of IDs for all messages in the group that are older than the coalesced record.
+                # This is our snapshot of messages to process (delete) in this coalesced group.
+                all_ids = list(
+                    self.select_coalesced_messages()
+                    .filter(id__lt=coalesced_id)
+                    .values_list("id", flat=True)
+                    .order_by("id")
+                )
+
+                # Reserve the messages for processing. By updating scheduled_for to a point
+                # in the future (now + 1 hour), we effectively signal that these messages are
+                # being processed, thereby preventing any other drain process from picking them up.
+                self.objects.filter(id__in=all_ids + [coalesced_id]).update(
+                    scheduled_for=timezone.now() + datetime.timedelta(hours=1)
+                )
```

Comment:
Moving to a reservation update style approach could help reduce the statement timeouts we encounter currently on long running outbox signal handlers as we're no longer running them inside the transaction. 🤔